pkg/storegateway/indexcache/cache.go

```diff
@@ -91,23 +92,22 @@ type PostingsKey string
 
 // CanonicalPostingsKey creates a canonical version of PostingsKey
 func CanonicalPostingsKey(postings []storage.SeriesRef) PostingsKey {
-	hashable := make([]byte, len(postings)*8)
-	for i, posting := range postings {
-		for octet := 0; octet < 8; octet++ {
-			hashable[i+octet] = byte(posting >> (octet * 8))
-		}
-	}
-
-	// We hash the postings list twice to minimize the chance of collisions
-	hasher1, _ := blake2b.New256(nil) // This will never return an error
-	hasher2 := sha1.New()
-
-	_, _ = hasher1.Write(hashable)
-	_, _ = hasher2.Write(hashable)
-
-	checksum := hasher2.Sum(hasher1.Sum(nil))
+	hashable := unsafeCastPostingsToBytes(postings)
+	checksum := blake2b.Sum256(hashable)
+	return PostingsKey(base64.RawURLEncoding.EncodeToString(checksum[:]))
+}
 
-	return PostingsKey(base64.RawURLEncoding.EncodeToString(checksum))
+const bytesPerPosting = int(unsafe.Sizeof(storage.SeriesRef(0)))
+
+// unsafeCastPostingsToBytes returns the postings as a slice of bytes with minimal allocations.
+// It casts the memory region of the underlying array to a slice of bytes.
+func unsafeCastPostingsToBytes(postings []storage.SeriesRef) []byte {
+	byteSlice := make([]byte, 0)
+	slicePtr := (*reflect.SliceHeader)(unsafe.Pointer(&byteSlice))
```

Comment:
non-blocking: Could we add something that's checked at compile time to ensure if the layout here ever changes we'll find out about it before a runtime error? I don't know exactly what this would look like.

Reply:
pasting from a DM - there are tests to make sure that using this unsafe hack is the same as doing the safe thing. If we change that postings aren’t castable to bytes, then [this helper function](https://github.com/grafana/mimir/pull/3779/files#diff-88cac663744a42cfb381dc521559f6c21b4afa7ce57cbd82246ee4adc52e35f4R62-R70) in the tests won’t compile